### PR TITLE
feat: Model Transformation Layer — Native PipelineStage modifiers (#410)

### LINKED ISSUE #410: Model Transformation Layer - Post-loading modifications (freezing, transfer learning, adapters)
## Overview
During my work towards #248 I have converged on a solution that requires a refactor of the model initialisation.

**UPDATE**: This issue is now part of the 5-PR checkpoint architecture refactor. It implements the **Model Transformation Layer** that applies post-loading modifications to models.

## Parent Issue
Part of #248 - Checkpoint System Refactor

## Related Architecture Components
- Foundation: #493 (Pipeline Infrastructure)
- Works with: #458 (Checkpoint Acquisition), #494 (Loading Orchestration)
- Integration: #495 (System Integration)

## Original Problem
Specifically in https://github.com/ecmwf/anemoi-core/blob/d51219f0ef598a149dc2c7901b055947785a26b2/training/src/anemoi/training/train/train.py#L217

We have a lot of branching that is getting increasingly difficult to extend and maintain.

## New Architecture Role
This issue implements the **Model Transformation Layer** - the third layer in our pipeline architecture:

```
┌─────────────────────────────────────────────────┐
│    Model Transformation Layer (THIS ISSUE)      │
│         (Post-loading modifications)            │
├─────────────────────────────────────────────────┤
│         Loading Orchestration Layer             │
│    (Strategies for applying checkpoints)        │
├─────────────────────────────────────────────────┤
│        Checkpoint Acquisition Layer             │
│      (Obtaining checkpoint from sources)        │
└─────────────────────────────────────────────────┘
```

### 1. ModelModifier Base Class (`training/src/anemoi/training/train/modify.py`)
```python
class ModelModifier(ABC):
    """Base class for all model modifiers"""
    @abstractmethod
    def apply(self, model: nn.Module) -> nn.Module:
        """Apply modification to model"""
        pass
```

### 2. Implemented Modifiers
- `FreezingModelModifier`: Selective parameter freezing by module name
- `TransferLearningModelModifier`: Transfer learning with fallback for missing checkpoint system
- Future: `LoRAModifier`, `QuantizationModifier`, etc.

### 3. ModelModifierApplier
Orchestrates application of multiple modifiers in sequence:
```python
class ModelModifierApplier:
    def process(self, base_model: nn.Module, config: DictConfig) -> nn.Module
```

## Configuration
```yaml
training:
  model_modifier:
    modifiers:
      - _target_: "anemoi.training.train.modify.FreezingModelModifier"
        submodules_to_freeze:
          - "encoder"
          - "processor.0"
      - _target_: "anemoi.training.train.modify.TransferLearningModelModifier"
        checkpoint_path: "/path/to/pretrained.ckpt"
        strict: false
        skip_mismatched: true
```

## Benefits
- **Modular**: Each modification type is a separate, composable modifier
- **Configurable**: Full Hydra configuration support with validation
- **Extensible**: Easy addition of new modifier types
- **Standalone**: Works independently without external dependencies
- **Order-aware**: Modifiers applied in specified sequence

## Implementation Status (PR #442)
- [x] Base ModelModifier class
- [x] FreezingModelModifier
- [x] TransferLearningModelModifier
- [x] ModelModifierApplier
- [x] Configuration integration
- [ ] Tests
- [ ] Documentation

## Future Extensions
This modular system enables:
- PEFT adapters (LoRA, QLoRA)
- Quantization
- On-the-fly renaming (#249)
- Conditional loading into modified architectures
- Pre-training on one decoder, fine-tuning on multi-decoder

## Testing
- Unit tests for each modifier
- Integration tests with training pipeline
- Configuration validation tests

## Success Criteria
- [ ] All existing functionality preserved
- [ ] Clean separation of concerns
- [ ] Extensible for future modifiers
- [ ] Full test coverage
- [ ] Documentation complete

### PR DESCRIPTION
**Pull request by JesperDramsch**
**Part of the checkpoint system refactor (#248).** Implements the Model Transformation Layer as native `PipelineStage` classes built on Phase 1 infrastructure (PR #501).

> ⚠️ **Architecture change from original PR.** The original `train/modify.py` (sync `ModelModifier.apply(model)` interface) has been **deleted and replaced** with native async `PipelineStage` implementations in `checkpoint/modifiers/`. See design decisions below.

### What this PR does

- **`FreezingModifierStage`** — native `PipelineStage` that freezes specified submodules. Full feature port from the original `FreezingModelModifier`: strict mode, gradient validation, parameter counting, dot-notation submodule lookup.
- **Deletes `train/modify.py`** — `ModelModifierApplier`, `FreezingModelModifier`, and `TransferLearningModelModifier` are all removed. Transfer learning now lives in the loading layer (#998). Pipeline handles orchestration.
- **Config templates** for all 4 loading strategies (`weights_only.yaml`, `transfer_learning.yaml`, `warm_start.yaml`, `cold_start.yaml`)
- **Package exports** and catalog discovery for the modifier layer

### Architecture

```
CheckpointContext enters with model (weights already loaded by loading strategy)
        ↓
  FreezingModifierStage freezes specified submodules
        ↓
CheckpointContext exits with model modified + metadata["modifiers_applied"] updated
```

Composes naturally with sources and loaders:
```python
pipeline = CheckpointPipeline(stages=[
    LocalSource(),
    TransferLearningLoader(skip_mismatched=True),
    FreezingModifierStage(submodules_to_freeze=["encoder"]),
])
```

### Why `train/modify.py` was deleted (not wrapped)

The original plan was thin adapter wrappers around the existing sync classes. That turned out to be wrong:

- PR #410's code was prototype, not production-ready
- Wrapping sync `apply(model)` inside async `process(context)` creates dual code paths
- `ModelModifierApplier` duplicated pipeline orchestration that Phase 1 already handles
- `TransferLearningModelModifier` mixed loading + modification concerns — these are now separate layers

Building native `PipelineStage` classes is cleaner: no indirection, single code path, clear separation of concerns.

### Current status

🟢 **Ready for review.** `main` is merged in, so the diff is now scoped to the modifier delta.

- ✅ g1: FreezingModifierStage native port + train/modify.py deleted
- ✅ g2: Config templates
- ✅ g3: Package exports + catalog discovery (ModelModifier ABC + catalog registration)
- ✅ g4: Modifier + discovery tests green (242 checkpoint unit tests pass)
- 🟢 g5: Submitted for review

The catalog-discovery suite also checks that the loading strategies import; that
single check depends on #998 landing on `main` and is out of scope for this PR.

### What's NOT in this PR

| Deferred | Why | Tracked in |
|----------|-----|-----------|
| PEFT/LoRA modifiers | Phase 3 roadmap | #248 |
| QLoRA, Prefix Tuning | Phase 3 roadmap | #248 |
| `train.py` wiring | Phase 3 integration | #495 |

### Related PRs

- **#501** — Phase 1 Pipeline Infrastructure (merged, dependency)
- **#464** — Checkpoint Sources (in review)
- **#998** — Loading Orchestration (in review)
- **#495** — Phase 3 Integration (blocked on Phase 2)

**Part of #248**

<details>
<summary>Original PR description (archived 2026-03-20)</summary>

## Overview
During my work towards #248 I have converged on a solution that requires a refactor of the model initialisation.

This issue implements the **Model Transformation Layer** that applies post-loading modifications to models.

### Components
- `FreezingModelModifier`: Selective parameter freezing by module name
- `TransferLearningModelModifier`: Transfer learning with fallback for missing checkpoint system
- `ModelModifierApplier`: Orchestrates application of multiple modifiers in sequence

### Configuration
```yaml
training:
  model_modifier:
    modifiers:
      - _target_: "anemoi.training.train.modify.FreezingModelModifier"
        submodules_to_freeze:
          - "encoder"
          - "processor.0"
```

</details>

----
📚 Documentation preview 📚: https://anemoi-training--442.org.readthedocs.build/en/442/

----
📚 Documentation preview 📚: https://anemoi-graphs--442.org.readthedocs.build/en/442/

----
📚 Documentation preview 📚: https://anemoi-models--442.org.readthedocs.build/en/442/

----
📚 Documentation preview 📚: https://anemoi-training--442.org.readthedocs.build/en/442/

----
📚 Documentation preview 📚: https://anemoi-graphs--442.org.readthedocs.build/en/442/

----
📚 Documentation preview 📚: https://anemoi-models--442.org.readthedocs.build/en/442/

### CODE REVIEW

**Comment by JPXKQX** [comment]
Thanks Jesper, I think this refactoring makes a lot of sense. Would it make sense to have another modifier "ResumeRun..." to which we could bring all the logic from  `run_id`, `fork_run_id`, `load_only_weights` and `warm_start`?

**Comment by JesperDramsch** [comment]
> Thanks Jesper, I think this refactoring makes a lot of sense. Would it make sense to have another modifier "ResumeRun..." to which we could bring all the logic from `run_id`, `fork_run_id`, `load_only_weights` and `warm_start`?

Possibly. I believe my original design around `fork_run_id` ended up confusing most people, so we could take a look whether this design could be used to fix that.

File: training/docs/user-guide/training.rst [hunk 1]
```diff
@@ -562,39 +562,55 @@ The above can be adapted depending on the use case and taking advantage
...
+
+Parameter freezing with multiple layers:
+
+.. code:: yaml
+
+   training:
+      model_modifier:
+         modifiers:
+            - _target_: "anemoi.training.checkpoint.modifiers.freezing.FreezingModifierStage"
+              submodules_to_freeze: ["encoder", "processor.0"]
```
Comment (anaprietonem) [inline training/docs/user-guide/training.rst]: @JPXKQX since you mentioned that you recently spot an issue with freezing, can you confirm this fits fine with the test you were doing in the multi-dataset set up?

File: training/src/anemoi/training/checkpoint/modifiers/freezing.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        """
+        try:
+            target_module = module.get_submodule(target_name)
+        except AttributeError:
+            return None
+
+        frozen_count = 0
+        for param in target_module.parameters():
+            if param.requires_grad:
+                param.requires_grad = False
```
Comment (anaprietonem) [inline training/src/anemoi/training/checkpoint/modifiers/freezing.py]: guess you could also do here something like?
```
frozen_count = sum(p.requires_grad for p in target_module.parameters())
target_module.requires_grad_(False)
```

File: training/src/anemoi/training/checkpoint/modifiers/freezing.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                frozen_count += 1
+        return frozen_count
+
+    def _validate_gradient_flow(self, model: torch.nn.Module) -> None:
```
Comment (anaprietonem) [inline training/src/anemoi/training/checkpoint/modifiers/freezing.py:139]: do you have a feeling for how expensive is to do this check?

File: training/src/anemoi/training/config/training/single.yaml [hunk 4]
```diff
@@ -4,6 +4,7 @@ defaults:
   - training_loss: single
   - optimization: default
   - weight_averaging: null
+  - optional model_modifier: null
```
Comment (anaprietonem) [inline training/src/anemoi/training/config/training/single.yaml]: do we need this now? As in if the integration work will come after we could just leave the configs and then update all the other top level configs in the final integration PR

File: training/docs/user-guide/training.rst [hunk 5]
```diff
@@ -562,39 +562,55 @@ The above can be adapted depending on the use case and taking advantage
 of hydra, you can also reuse ``config.training.run_id`` or
 ``config.training.fork_run_id`` to define the path to the checkpoint.
 
-*******************
- Transfer Learning
```
Comment (anaprietonem) [inline training/docs/user-guide/training.rst:566]: Don't think the section about transfer learning should be removed.

File: training/docs/user-guide/training.rst [hunk 6]
```diff
@@ -562,39 +562,55 @@ The above can be adapted depending on the use case and taking advantage
...
+
+   training:
+      model_modifier:
+         modifiers:
+            - _target_: "anemoi.training.checkpoint.modifiers.freezing.FreezingModifierStage"
+              submodules_to_freeze: ["encoder", "processor.0"]
+              strict: false
+              validate_gradients: true
+
+Legacy key
```
Comment (anaprietonem) [inline training/docs/user-guide/training.rst]: Similarly, the comment below. At the moment the modifiers can't be used till we have the integration PR, so I think it would be better to change this page then so we simply update it and no need to keep any 'legacy entries'

**Review by anaprietonem — COMMENTED** [review]
Nice work @JesperDramsch ! thanks for the PR. Left some minor suggestions. Happy to have a another look after!

File: training/src/anemoi/training/config/training/single.yaml (hunk 4 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/config/training/single.yaml]: That's fair. I was leaning towards "keeping it moving" rather than wait. Pushing to Phase 3

File: training/docs/user-guide/training.rst (hunk 5 above)
Comment (JesperDramsch) [reply training/docs/user-guide/training.rst:566]: Agreed, reverted the user-guide back to `main` so the Transfer Learning section is restored.

File: training/docs/user-guide/training.rst (hunk 6 above)
Comment (JesperDramsch) [reply training/docs/user-guide/training.rst]: Yep, makes sense. Dropped the modifier docs here, I'll redo this page in the integration PR once they're actually usable. No legacy entries.

File: training/src/anemoi/training/checkpoint/modifiers/freezing.py (hunk 2 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/checkpoint/modifiers/freezing.py]: Good point, done. Swapped to `sum(p.requires_grad ...)` + `requires_grad_(False)`, same behaviour just less code.

File: training/src/anemoi/training/checkpoint/modifiers/freezing.py (hunk 3 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/checkpoint/modifiers/freezing.py:139]: Honestly negligible. It was a one-off forward/backward at load, but the dummy `(1,10)` input doesn't match real models so it just warned and skipped there anyway. Swapped it for a plain `requires_grad` check on the frozen submodules, which is free, and `requires_grad=False` already guarantees no grads.

**Comment by JesperDramsch** [comment]
Thanks for the review @anaprietonem I believe I addressed all comments.

**Review by anaprietonem — APPROVED** [review]
Thanks for the contribution Jesper! and for the addressing the comments, good to be merged!
